File: src-tauri/src/novel_features.rs
//! 小说功能选项：题材 / 核心玩法 / 风格 / 关系 / 受众（写作与世界观生成用中文标签）。
use crate::models::NovelFeatures;

fn label(id: &str) -> &str {
    match id {
        "urban" => "都市",
        "xuanhuan" => "玄幻",
        "xianxia" => "仙侠",
        "scifi" => "科幻",
        "western_fantasy" => "西幻",
        "history" => "历史",
        "military" => "军事",
        "mystery" => "悬疑",
        "supernatural" => "灵异",
        "game" => "游戏",
        "apocalypse" => "末世",
        "system" => "系统",
        "rebirth" => "重生",
        "transmigration" => "穿越",
        "infinite_flow" => "无限流",
        "farming" => "种田经营",
        "lord_building" => "领主建设",
        "tycoon" => "神豪",
        "cautious" => "苟道",
        "academy" => "学院",
        "rule_horror" => "规则怪谈",
        "dungeon" => "副本挑战",
        "many_children" => "多子多福",
        "cool" => "爽文",
        "funny" => "搞笑",
        "hotblood" => "热血",
        "dark" => "黑暗",
        "light" => "轻松",
        "healing" => "治愈",
        "horror" => "恐怖",
        "realistic" => "现实",
        "ensemble" => "群像",
        "epic" => "史诗",
        "single_heroine" => "单女主",
        "multi_heroine" => "多女主",
        "no_cp" => "无CP",
        "dual_heroine" => "双女主",
        "harem" => "后宫",
        "pure_love" => "纯爱",
        "male" => "男频",
        "female" => "女频",
        other => other,
    }
}

fn join_labels(ids: &[String]) -> String {
    ids.iter()
        .map(|s| label(s.trim()))
        .filter(|s| !s.is_empty())
        .collect::<Vec<_>>()
        .join("、")
}

pub fn features_is_empty(f: &NovelFeatures) -> bool {
    f.genres.is_empty()
        && f.core_play.is_empty()
        && f.styles.is_empty()
        && f.relationships.is_empty()
        && f.audiences.is_empty()
}

/// 世界观 / 写作 Prompt 用的功能选项块（中文）。
pub fn format_novel_features_block(f: &NovelFeatures) -> String {
    if features_is_empty(f) {
        return "（未配置功能选项）".into();
    }
    let mut lines = Vec::new();
    if !f.genres.is_empty() {
        lines.push(format!("题材：{}", join_labels(&f.genres)));
    }
    if !f.core_play.is_empty() {
        lines.push(format!("核心玩法：{}", join_labels(&f.core_play)));
    }
    if !f.styles.is_empty() {
        lines.push(format!("风格：{}", join_labels(&f.styles)));
    }
    if !f.relationships.is_empty() {
        lines.push(format!("关系：{}", join_labels(&f.relationships)));
    }
    if !f.audiences.is_empty() {
        lines.push(format!("受众：{}", join_labels(&f.audiences)));
    }
    lines.join("\n")
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn formats_preset_and_custom() {
        let f = NovelFeatures {
            genres: vec!["urban".into(), "赛博修真".into()],
            core_play: vec!["system".into()],
            styles: vec![],
            relationships: vec!["no_cp".into()],
            audiences: vec!["male".into()],
        };
        let s = format_novel_features_block(&f);
        assert!(s.contains("题材：都市、赛博修真"));
        assert!(s.contains("核心玩法：系统"));
        assert!(s.contains("关系：无CP"));
        assert!(s.contains("受众：男频"));
        assert!(!s.contains("风格"));
    }
}

File: src/lib/novelFeatures.ts
/** Novel positioning tags (create form + root panel). Preset ids are stable; custom = free text. */

export type NovelFeatures = {
  genres: string[];
  core_play: string[];
  styles: string[];
  relationships: string[];
  audiences: string[];
};

export type FeatureGroupKey = keyof NovelFeatures;

export type FeatureGroupDef = {
  key: FeatureGroupKey;
  /** i18n: novels.feat.group.{key} */
  allowCustom: boolean;
  options: readonly string[];
};

export const FEATURE_GROUPS: readonly FeatureGroupDef[] = [
  {
    key: "genres",
    allowCustom: true,
    options: [
      "urban",
      "xuanhuan",
      "xianxia",
      "scifi",
      "western_fantasy",
      "history",
      "military",
      "mystery",
      "supernatural",
      "game",
      "apocalypse",
    ],
  },
  {
    key: "core_play",
    allowCustom: true,
    options: [
      "system",
      "rebirth",
      "transmigration",
      "infinite_flow",
      "farming",
      "lord_building",
      "tycoon",
      "cautious",
      "academy",
      "rule_horror",
      "dungeon",
      "many_children",
    ],
  },
  {
    key: "styles",
    allowCustom: true,
    options: [
      "cool",
      "funny",
      "hotblood",
      "dark",
      "light",
      "healing",
      "horror",
      "realistic",
      "ensemble",
      "epic",
    ],
  },
  {
    key: "relationships",
    allowCustom: false,
    options: ["single_heroine", "multi_heroine", "no_cp", "dual_heroine", "harem", "pure_love"],
  },
  {
    key: "audiences",
    allowCustom: false,
    options: ["male", "female"],
  },
] as const;

export function emptyNovelFeatures(): NovelFeatures {
  return {
    genres: [],
    core_play: [],
    styles: [],
    relationships: [],
    audiences: [],
  };
}

export function normalizeNovelFeatures(raw: unknown): NovelFeatures {
  const base = emptyNovelFeatures();
  if (!raw || typeof raw !== "object") return base;
  const o = raw as Record<string, unknown>;
  for (const g of FEATURE_GROUPS) {
    const v = o[g.key];
    if (!Array.isArray(v)) continue;
    const preset = new Set(g.options);
    const out: string[] = [];
    const seen = new Set<string>();
    for (const item of v) {
      if (typeof item !== "string") continue;
      const s = item.trim();
      if (!s || seen.has(s)) continue;
      if (preset.has(s) || g.allowCustom) {
        seen.add(s);
        out.push(s);
      }
    }
    base[g.key] = out;
  }
  return base;
}

export function toggleFeatureValue(
  features: NovelFeatures,
  group: FeatureGroupKey,
  value: string,
): NovelFeatures {
  const cur = features[group];
  const next = cur.includes(value) ? cur.filter((x) => x !== value) : [...cur, value];
  return { ...features, [group]: next };
}

export function addCustomFeature(
  features: NovelFeatures,
  group: FeatureGroupKey,
  raw: string,
): NovelFeatures {
  const def = FEATURE_GROUPS.find((g) => g.key === group);
  if (!def?.allowCustom) return features;
  const s = raw.trim();
  if (!s) return features;
  if (features[group].includes(s)) return features;
  return { ...features, [group]: [...features[group], s] };
}

export function isPresetOption(group: FeatureGroupKey, value: string): boolean {
  const def = FEATURE_GROUPS.find((g) => g.key === group);
  return !!def?.options.includes(value);
}

export function customValues(features: NovelFeatures, group: FeatureGroupKey): string[] {
  const def = FEATURE_GROUPS.find((g) => g.key === group);
  if (!def) return [];
  const preset = new Set(def.options);
  return features[group].filter((v) => !preset.has(v));
}
